addon/components/gavant-table.ts

```diff
@@ -0,0 +1,532 @@
+                let isVisible = (col.isFixedLeft && allowFixedCols) || newTableWidth + colWidth <= containerWidth;
+                if (isVisible) {
+                    newTableWidth += colWidth;
+                    visibleColumns.pushObject(col);
+                    // Prevent unwanted panning behavior on mobile that occurs with more than 2 columns
+                    const isMobile = get(get(this, 'media'), 'isMobile');
+                    if (isMobile && visibleColumns.length === 2) {
+                        break;
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+
+        setProperties(this, { containerWidth, visibleColumns });
+        this.changeDisplay();
+    }
+
+    /**
+     * Pans the table's visible column to the left or right the provided number of columns
+     * e.g. `-1` pans one column to the left, `2` pans two columns to the right
+     *
+     * @param {number} moveIndex -- A number indicating the number of columns to pan
+     * @memberof GavantTableComponent
+     */
+    panColumns(moveIndex: number) {
+        const newColumnPanPosition = this.columnPanPosition + moveIndex;
+        const pannedColumns = this.allowFixedCols ? this.nonFixedColumns : this.columns;
+        if (newColumnPanPosition < 0 || newColumnPanPosition >= pannedColumns.length) {
+            return;
+        }
+
+        set(this, 'columnPanPosition', newColumnPanPosition);
+        this.updateColumnVisibility();
+    }
+
+    /**
+     * Returns the computed width for the given element
+     *
+     * @param {Element} el -- A DOM element
+     * @returns {number}
+     * @memberof GavantTableComponent
+     */
+    getElementWidth(el: Element): number {
+        let width = 0;
+
+        if (el) {
+            let rects = el.getClientRects();
+            let paddingLeft = parseFloat(window.getComputedStyle(el, null).getPropertyValue('padding-left'));
+            let paddingRight = parseFloat(window.getComputedStyle(el, null).getPropertyValue('padding-right'));
+            if (rects) {
+                width = paddingLeft && paddingRight ? rects[0].width - (paddingLeft + paddingRight) : rects[0].width;
+            }
+        }
+
+        return width;
+    }
+
+    /**
+     * Returns the column's unique identifier or undefined
+     *
+     * @param {(TableColumn | undefined)} col - A table column
+     * @returns {(string | undefined)}
+     * @memberof GavantTableComponent
+     */
+    getColumnId(col: TableColumn | undefined): string | undefined {
+        //allow columns to specify an "id" if they dont have a valuePath, or the valuePath is not unique
+        return col && (col.id || col.valuePath);
+    }
+
+    //actions
+
+    /**
+     * Invokes an action to load a new page of rows
+     *
+     * @returns {Promise}
+     * @memberof GavantTableComponent
+     */
+    @action
+    onLastReached() {
+        if (!this.isLoading && this.hasMoreRows) {
```

Comment:
I wouldn't use `tryInvoke` as Ember is deprecating most of the `try` helpers and it sorta ruins the type safety.

Right now your comments on this action are possibly incorrect, because loadMoreRows might not exist and therefore it returns the function that tryInvoke returns (not a Promise). I think you will want to check if `this.loadMoreRows` exists in the if check, and then just call it

Reply:
Makes sense. Since this will require declaring a property called `loadMoreRows`, what's the preferred type for functions?

Reply:
@napafundi depends on what params the function accept, and what its return type is, but in general you'll define function types like this:
```ts
someFunc: (param1: type, param2: anotherType) => returnType;
```
in this case, I think this action takes no params and returns a promise, so probably something like:
```ts
loadMoreRows?: () => Promise<any[]>;
```
which means an optional function with no arguments that returns a promise (if it returned nothing, that would be `void` instead of `Promise`) which resolves with an array of objects in any shape (that could possibly be more specific, but the table doesn't really care about the row shapes, so I think thats probably ok in this case).